### ISSUE TITLE
build fails - TypeError: can't apply this __setattr__ to DefaultMeta object

### ISSUE THREAD

**Issue by rabble** [lucyparsons/OpenOversight#758]
Code: 3b32dea

Make dev runs for a while and gives this. I'm not a big python person.... is it a version problem?

`Successfully built b22144b72e7b
Successfully tagged openoversight_web:latest
docker-compose up -d
openoversight_postgres_1 is up-to-date
Recreating openoversight_web_1 ... done
Postgres is up
## Creating database
docker-compose run --rm web /usr/local/bin/python ../create_db.py
Starting openoversight_postgres_1 ... done
Traceback (most recent call last):
  File "../create_db.py", line 2, in <module>
    from OpenOversight.app import create_app
  File "/usr/src/app/OpenOversight/app/__init__.py", line 103, in <module>
    app = create_app()
  File "/usr/src/app/OpenOversight/app/__init__.py", line 32, in create_app
    from .models import db  # noqa
  File "/usr/src/app/OpenOversight/app/models.py", line 18, in <module>
    db = SQLAlchemy()
  File "/usr/local/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 677, in __init__
    self.Model = self.make_declarative_base(model_class, metadata)
  File "/usr/local/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 758, in make_declarative_base
    model.query_class = self.Query
  File "/usr/local/lib/python3.8/site-packages/sqlalchemy/ext/declarative/api.py", line 68, in __setattr__
    _add_attribute(cls, key, value)
  File "/usr/local/lib/python3.8/site-packages/sqlalchemy/ext/declarative/base.py", line 639, in _add_attribute
    type.__setattr__(cls, key, value)
TypeError: can't apply this __setattr__ to DefaultMeta object
make: *** [create_db] Error 1`

**Comment by r4v5** [comment]
@rabble What you're running into should have been fixed by 3b32dea. You may need to destroy and rebuild the container. The tell is the lines `/usr/local/lib/python3.8` in the stacktraces; 3b32dea should have made it build with python3.5 unless you explicitly set TRAVIS_PYTHON_VERSION to something else. (the python 3.8 base container currently has a breaking bug with sqlalchemy.)

Docker-compose can be kinda stubborn about building containers. We have a `make` task to delete prebuilt image containers (`make clean`) and one for building new ones (`make build`) that hopefully will help correct the container setup.

If you're still having troubles with the tooling you can build the image manually:
```
openoversight on develop [?] on ☁️  us-east-1 
✦ ❯ export UID ; docker-compose down
WARNING: The AWS_ACCESS_KEY_ID variable is not set. Defaulting to a blank string.
WARNING: The AWS_SECRET_ACCESS_KEY variable is not set. Defaulting to a blank string.
WARNING: The AWS_DEFAULT_REGION variable is not set. Defaulting to a blank string.
WARNING: The S3_BUCKET_NAME variable is not set. Defaulting to a blank string.
WARNING: The APPROVE_REGISTRATIONS variable is not set. Defaulting to a blank string.
Stopping openoversight_web_1      ... done
Stopping openoversight_postgres_1 ... done
Removing openoversight_web_1      ... done
Removing openoversight_postgres_1 ... done
Removing network openoversight_default

openoversight on develop [?] on ☁️  us-east-1 took 12s 
✦ ❯ docker rmi openoversight_web    
Untagged: openoversight_web:latest
Deleted: sha256:c85955135a17ec9078b9fe871cf4a97d4fee3e8501c1febc5a581ac0462e9ccc
Deleted: sha256:ecbcd9236a2c17ab957d52a18cdd7df72b5cf88e675dbdabc0096f8cad5b6687
Deleted: sha256:b1f81db58276c7169765788346f76fb5edf6f3e7eebfb7db2b70167fdb6bfe11
Deleted: sha256:76dc0b36c0324aa7e943c2fe8776d5f9b8985a6a18625d398e0d180a1bbf858f
Deleted: sha256:0602aa81703256ad39666801b747358f76c88fdbc579dd8add5aa8ff23d2691f
Deleted: sha256:86b9b4e36ffa5591700201f2152372c54b5ebac32a13bf00823d9e53964b4e5b
Deleted: sha256:b453e2c3e154a1ff8c9c2ab5c7e257f8560921af2e6adb428252711a3a07032e
Deleted: sha256:49acc01d172a524811a5ddca384b0e1a3216cbcfdb4bbd434c6c1e32800b15a8
Deleted: sha256:e31dfb01dd0e839bcabf969ce5030937dce1eb3ee66680c1f079ab5843a74017
Deleted: sha256:e23661e214c1cd4e50305c98cee6f178cee6b8d6ddae4381025c62625e08ce94
Deleted: sha256:84408896b685171639f45c3517de48b3bd0267800b839bc151cca77a7a3120cd
Deleted: sha256:e19bac0a5cf0aaf400304a2ff7ca24d48d3dba7db101390d7fa93f6043da7ac6
Deleted: sha256:ced75258dc9f606fa2955b8ea5059f1fd1958124fd3bab4bc13e31cf55fc9193
Deleted: sha256:aade96da6ab0ac5c14c934d2f04d2c51c2d1b6abdb3e61cc500836edc923c8f7
Deleted: sha256:fbd8957934e170f3e7f5d3ebaa7bf4672dad8236a0a5f4f8a20f3d4641f837da
Deleted: sha256:697d38e2d2c0753dfebde41eef05820e7ea68109148f0e885f5af6ca6437c50b
Deleted: sha256:333d23b7e23bd91000b27f9ab4ac46a9fd6f3d331e820bce38d1769633666e57
Deleted: sha256:19241d53a5223f8d31684d04bf5d21f89f8e4bd1223160c58cacd6382c63401c
Deleted: sha256:f01a2b5433319353cb68eb3f8c660e147d23889bbcc320a0c59c1be0786edbf0
Deleted: sha256:b7b4c4c162beab343c6fedbe7367bc37ea1376d10ee52254f5dac1ccb3c4893b
Deleted: sha256:bec919fe5386b84c030516982d3751d242436b0bc2ec9781703d495a0cb5dadb
Deleted: sha256:46c2e1e3ba300b027a0be3804d10dc0d6de84c60f5336a8aa22be2cbef427440
Deleted: sha256:c96e06ed55d3c40e97ea73969fae20e16a741ed2799be6c5b6c6ed880b69b759
Deleted: sha256:1b38dca9a251f22c10bf8273a1b1d24605fdca6faa592127312dfe5b6c247fb1
Deleted: sha256:602b211bccc662b52d2f5152149645db6714ce830d7b28b2674d4999c512a5d4
Deleted: sha256:756a9e11dda4c2bc793e9f1d534b1fa4b060df5186d6bdde31563cdc19b0a5e2
Deleted: sha256:b222f44d7ce59e25666197298c6011a46034ef0842ece760791b93a61825da71
Deleted: sha256:2fa6a436bb65aaf5e8e283d597c14318423cb0425f21f498a7bbe303e2a21e64
Deleted: sha256:cbe266aaed1cc1be21ef7755dbd352ddf9de50df77cd7aa867b7ca3b8849e892
Deleted: sha256:e03bd3d4dcc3ed4846e6591c8a1602b7380c02b67755abf71571a6bcdb6a3f12
Deleted: sha256:dff534f5b3e56a53f6e792dafe9d6ace4a2a25095701c5c43127d497bd8e3c74
Deleted: sha256:9e5fd315eefe8da4268e2d272a2d0cce56d010441639ca5b7d4491aaa3ca7faa
Deleted: sha256:7c4a5e0848bf17cf4e875a708259b3388d0a66fd47b249d17e79994b0f718a45
Deleted: sha256:f9dcdbe87810ad8cd8c099028a4251110bd359188645dd78815087b0bc1a9de2
Deleted: sha256:e5a430fe517609723169242f4ef99d5d541dd8af0c6a69cae7b65c8be046cfe9
Deleted: sha256:21cfea190fb50d89826df793dac551a8792962088890257e2cec1890adbdff15
Deleted: sha256:12953294974ef21caca1d8c316304565037c998ce169fd55b6540255f0201450
Deleted: sha256:46303af6162e2024162b73f18a8852fa4fb1ca7c7da3912a3b857a3120b97774
Deleted: sha256:0b5ebc486cac5c0a7d96fd98b47a77b0701c82dd6c5da3eab547db3e06c28816
Deleted: sha256:4c47fd164940c9922d1503bf70b02c3b8e90563dda925a8e60e304ea99734748

openoversight on develop [?] on ☁️  us-east-1 
✦ ❯ docker build -t openoversight_web -f dockerfiles/web/Dockerfile .
Sending build context to Docker daemon  38.69MB
Step 1/30 : ARG TRAVIS_PYTHON_VERSION
Step 2/30 : FROM python:${TRAVIS_PYTHON_VERSION:-3.5}-buster
 ---> 85c44ea592e4
```
From there, docker-compose should find the right things.

If you're having trouble after that, reach out and I'll hop on a screenshare :)

(also, you may find the markdown syntax of three backticks on its own line to start and end a fixed-width text snippet, which is what I used up there, to be helpful when pasting python stacktraces -- they've got lots of double-underscores and asterisks, which tend to be interpreted as markdown formatting)

**Comment by am-i-the-rabbit** [comment]
@r4v5 Thanks for that!

I don't know a lot about SQLAlchemy, but is the issue with Python 3.8 a universal incompatibility between the two, or is it isolated to this particular project? I tried to find info on Google about any known conflicts between the two, but skimming the search results didn't appear to yield anything useful.

If this is an issue specific to this project, I'd like to figure out how to fix it, if that's even possible. I don't like the idea but how difficult would it be if we need to replace SQLAlchemy with a different SQL toolkit like sqldb or sqlite? Or is there a particular reason (other than being friendly with Flask) SQLAlchemy has to be used? 

The reason I ask is primarily in the interest of using at least Python 3.6, but if that's just a no-go, and SQLAlchemy needs to be used, I'm going to dig myself into a rabbit hole. XD

Forgive any of my ignorance, but I'm still acquainting myself with this project. I only just jumped in last night and I'm relatively new to DevOps. I'm a SysAdmin with some Python dev experience, but a lot of this is new to me. This just happens to be a cause near and dear to my heart, and I'm taking the opportunity to start building a foundation to shift into DevOps while I'm at it.

**Comment by r4v5** [comment]
@am-i-the-rabbit 

This bug is a follow-on from #754, which has slightly more detail about the problem; it's specifically [a regression in 3.8.4](https://bugs.python.org/issue41295)'s behavior that [hits a particular bit of flask-sqlalchemy](https://github.com/pallets/flask-sqlalchemy/issues/852). Because the image used for local development prior to 3b32dea was based on the '3.8-buster' docker label, it got silently updated from 3.8.3 to 3.8.4, which is how the regression ended up affecting us.

It looks like a library-side fix is also present in flask-sqlalchemy 2.4.4 (released july 14), so when we upgrade to that, we may also be able to build 3.8-buster images again.

> The reason I ask is primarily in the interest of using at least Python 3.6, but if that's just a no-go, and SQLAlchemy needs to be used, I'm going to dig myself into a rabbit hole. XD

Fortunately, there isn't a need to rip out flask-sqlalchemy just yet: both sqlalchemy and flask-sqlalchemy are mature and well-run projects (IMO), and I don't know that there would be value in the effort to move off either, especially given that the issue is resolved in the latest release.

The code in this project tries its best to work with every supported version of python, which as of right now is 3.5, 3.6, 3.7, and 3.8 (modulo the above regression :smile:). Travis runs tests on 3.5, 3.6, and 3.7 for every PR and every merged branch (and will have 3.8 readded shortly).

if you'd like to use 3.6 or 3.7 for local development or for building images for your own instance, you now can control that by setting the TRAVIS_PYTHON_VERSION environment variable, like `export TRAVIS_PYTHON_VERSION=3.6`, which will choose which version label gets used as a base image. Anything in https://hub.docker.com/_/python?tab=description that ends in '-buster' should work. I believe there is at least one instance of OO that operates using a deployment of the Docker image built from this repo, so they would have been running 3.8 this whole time (but have not rebuilt it since the image was updated to 3.8.4).

Because we want to be relatively easy to set up (for a python/js webapp, at least) and avoid the versioning hell that folks who aren't familiar with python/js application deployment tend to hit, we try to support as wide a range of Python versions as possible to meet people where their hosting provider is. We do have OO servers currently running 3.5, so the project absolutely must maintain 3.5 compatibility for now, which does mean missing out on some cool features like f-strings.

Python 3.5 [doesn't EOL until September](https://devguide.python.org/#status-of-python-branches) but when it does, we'll gladly drop support for it.

Re: moving from systems administration to devops: as someone who's done the same professionally, it's definitely a whole new set of challenges, but it feels great to no longer ever have to go down a hot aisle and I get to wear headphones instead of earplugs now.

**Comment by am-i-the-rabbit** [comment]
Right on! Thanks for all that.

I actually just ran across some comments a little while ago about that issue with regressing in 3.8.4. Good to know that it's not going to be a huge issue. It'll definitely give me something to do if I get bored. Hah!

Glad to hear I'm not the only one to make the jump to DevOps. Just seems like a lot more fun, honestly. I got interested in it last year when I started playing around on here for some WordPress extensions and just sort of fell in love with the whole process.

Thanks for all the info, and for the run down on some of the details for this. I definitely have a better idea, now, of what I'm looking at and...and for...

Many thanks, friend!
